### ISSUE TITLE
Splash screen not displayed for WinUI Desktop

### ISSUE THREAD

**Issue by Berkunath** [microsoft/microsoft-ui-xaml#4055]
**Splash screen not displayed for WinUI Desktop**
I have WinUI Desktop application and I would like to display the Splash screen before content of Window loads. But splash screen not displayed even we have Splash screen Images. Currently empty screen displays for 2 to 3 seconds.

**Expected behavior**
Splash Screen should display on start up of the application

**Version Info**
[Microsoft.WinUI 3.0.0-preview3.201113.0]

**Comment by StephenLPeters** [comment]
I think that Splash screens are a UWP concept. I think that if you want to show one in a win32 app, you'll need to do it yourself.  @Austin-Lamb might know more.

**Comment by marb2000** [comment]
@StephenLPeters is right. Splash Screens are not implemented yet in WinUI 3 Desktop. You can create a feature proposal :)

**Comment by jeganraj-m** [comment]
But there is an option to provide Splash screen in desktop projects.
![image](https://user-images.githubusercontent.com/74640452/111145182-fa4b3000-85ad-11eb-8c1f-d0f14c8ecbae.png)

Does it will be removed in future?

**Comment by AathifMahir** [comment]
Looking forward for native splash screen support added to WinUI 3, Splash Screen Like UWP does really makes a difference in User Experience

**Comment by CAKCy** [comment]
Why does it feel that with every new development platform Microsoft comes up with, things are going backwards? No splash screen, no XAML designer, right click text box crashing bug for more than a year, no topmost window etc. etc. What have we done to deserve Microsoft's "punishment"? :/

**Comment by michelkommers** [comment]
I will add my 'thoughts' on this, using  .Net MUAI with Blazor, splash screen becomes more important than before, because we have a lot of weird 'flashs' when the app is starting, we have some forecolor that flashs, than the white from the html page, than the 'background' of the page when the css loads, and only than we have the loading message inside the blazor... It may make sense in a dev perspective, but looking with the user point of vie....., its really bizare. With a splashscreen that loads all that in background and blazor signal back that everything is ready and the splashscreen fades out would be awesome.

**Comment by veikkoeeva** [comment]
Second to @michelkommers. For those coming later here from the linked Maui threads, I took a look at example provided by @castorix (thanks!), but at least for me it's not clear how to make use of it in a Maui app. Probably there's a way.

![kuva](https://user-images.githubusercontent.com/5158694/178119116-3d018b4f-394a-40b0-818b-ea14bce26380.png)

**Comment by Qws** [comment]
I made a MAUI app, the Android version has a default splashscreen, but when I click on the Windows version, nothing happens for 3-5 whole seconds. There is no feedback of the app loading, sometimes you click again, just to make sure the app is really loading. And sudenly the app pops up out of nowhere.

This really reminds me of the Windows XP user experience. I guess that's what I can expect from Win32 powered application...

But honestly, in modern time, it needs modern user experience. Not the Windows XP and Vista user experience.

Splashscreen should be the default option, not the otherway around... honestly, there is really no reason not to have a splashscreen for apps. 

Worst of all, UWP (WinUI2) has build-in splashscreen...... why would WinUI3 not have splashscreen. I feel like WinUI3 is going backwards in time.

**Comment by Laftek** [comment]
This is important!

**Comment by AathifMahir** [comment]
Exactly, Splash Screen is Must for WinUI 3 for better user experience and consistency sake, Currently many apps that uses WinUI 3 does implement custom splash screen that causes inconsistency between each different apps since many WinUI 3 Apps Uses Different Design and Different Implementation for Splash Screen.

In my case, on my DotNet Maui App. I developed a custom splash screen to complement the User experience by Using Same UWP Splash Screen Design Language but that differs from Other Implementations, Therefore having a native splash screen is a must on my case

**Comment by Laftek** [comment]
Best part is that Asignee is not even working for microsoft for some time…. Its sad how some fundamental functions are being neglected (also MAUI).

**Comment by breenbob** [comment]
Now working at Google... sad reflection of exodus from Microsoft. Also sad is that I love what they've done with Maui. And WinUI. All great steps in the right direction, IMO. But they need more attention and manpower. Bugs just take too damn long to get fixed. I guess they just don't bring in the same cash that Cloud computing does... 😞

**Comment by Laftek** [comment]
You are right, all kudos to the team MAUI and ASP.NET but as you mentioned lack of manpower is taking its price. I have big expectations for .NET8 tough! I think it will be gamechanger (kinda).

**Comment by Reza-Noei** [comment]
Any update on this ???
I don't wanna to re-invent the wheel.

**Comment by mklemarczyk** [comment]
We have end of 2023 soon, and WinUI does not support well SplashScreen or XAML Designer.
Any updates on the subject in this ticket?

**Comment by GuidoNeele** [comment]
Any updates? This issue is open for 3 years.

**Comment by AliShahesmaeili** [comment]
Any updates? This issue is open for 4 years.

**Comment by GuidoNeele** [comment]
I used [this](https://github.com/castorix/WinUI3_SplashScreen) by @castorix as inspiration to add our own splash screen.
We are using .NET MAUI on Windows and the startup time is super slow, you really need a splash screen to show the user something is happening.

**Comment by gmgallo** [comment]
Check [WinUIEx](https://github.com/dotMorten/WinUIEx) @dotMorten has a splash window for WinUI 3 in there and I think he may have something for MAUI.

**Comment by legistek** [comment]
I guess we'd better keep posting so the bot doesn't close it. 

After a full day of trial and error and a ridiculous amount of Win32 interop, I finally managed to make a splash screen that was properly centered, didn't flash white before being displayed, etc.

While I appreciate community efforts such as @dotMorten's, if this isn't part of the framework then I wanted to know how to do it myself. In fact successful community and self-help efforts only underscore how inexplicable it is to omit this from the framework. 

The splash screen is *literally* the first thing our users will see when they launch. (Ever hear that thing about first impressions?) Show me a Microsoft app (other than those built into Windows) that doesn't have a splash screen. For that matter show me a single retail Windows application by anyone that doesn't have a splash screen.

The lack of something this ubiquitous in the main framework is, to me, a sign of a bigger problem.

**Comment by NeerajGulia** [comment]
> After a full day of trial and error and a ridiculous amount of Win32 interop, I finally managed to make a splash screen that was properly centered, didn't flash white before being displayed, etc.

Would you like to share so that other people could also implement it.
Thanks in advance!

**Comment by legistek** [comment]
Oh I figured the one in [WinUIEx](https://github.com/dotMorten/WinUIEx) was as good if not better. I just wanted to learn for myself how to do it from scratch. It's quite a lot of code in different places and not really self-contained. It's basically how I'd have done it 30 years ago, except in C#. 

I will say one very tricky part was avoiding the white flash because the moronic default window will WM_PAINT itself white before it draws anything else **no matter what you do**. So I had to start with a transparent window using https://github.com/cnbluefire/WinUI3TransparentBackground which includes hooking into WndProc to override the moronic default behavior. But again I think WinUIEx is basically doing the same thing so I'd just look at his code and not my fingerpainting as the reference.

**Comment by jamers99** [comment]
@marb2000 it's been nearly 5 years. This experience is a significant downgrade from UWP. Is there anything that we can do to get our app up faster in an elegant way?

**Comment by dotMorten** [comment]
@jamers99 did you look at the new AoT feature? It specifically addresses start up time.
On a side note, a splash screen might make your app appear to start up faster (which could be beneficial for _percieved_ performance), but the actual app would only get slower, since it now has to do a splashscreen first.

**Comment by legistek** [comment]
> which could be beneficial for percieved performance

How about beneficial for perceived *professionalism*? Name me a single major commercial consumer-facing Windows application without a splash screen. There was a time when Microsoft cared about empowering its Windows developer community to make quality *professional* looking applications. But those days passed long ago.

**Comment by mklemarczyk** [comment]
> @jamers99 did you look at the new AoT feature? It specifically addresses start up time. On a side note, a splash screen might make your app appear to start up faster (which could be beneficial for _percieved_ performance), but the actual app would only get slower, since it now has to do a splashscreen first.

It will get only slower if you do some complex computation for splashscreen. As long I checked out to draw a frame on the screen it takes less than 1/60 of a second. Only extra time is loading the asset that can be optimized and restricted to avoid additional load on the application loading.

**Comment by mklemarczyk** [comment]
Any update with regards to roadmap for new year ?

**Comment by weezard12** [comment]
Any updates? This issue is open for 5 years.

**Comment by legistek** [comment]
Aren't we about due for WinUI3 to be deprecated and yet another newer, buggier, and less featureful Windows development framework to be introduced?

**Comment by dotMorten** [comment]
@weezard12 for what it’s worth WinUIEx can do it for you until the SDK supports it : https://dotmorten.github.io/WinUIEx/concepts/Splashscreen.html#simplesplashscreen

**Comment by mklemarczyk** [comment]
@weezard12 I would say the new update is to still use UWP. They just released support for .NET 9 in UWP and splash screen in "new" version works better than non existing one for WinUI3. I think UWP will stick for a little while if Microsoft want it or not, the gap for WinUI 3 is still far too big.

**Comment by weezard12** [comment]
@dotMorten @mklemarczyk Thank you for the advice,
I was able to figure out that I can use WinUIEx for the splash screen on Windows for my MAUI application.
Since it's such an old issue that many have struggled with, I wanted to know if there are other methods or updates with the topic.

I use WPF and MAUI but didn't heard a lot on UWP so I will check it out.
Again, thanks for the advice

**Comment by craigwi** [comment]
Here is another option for those who like the "old" UWP insitu splash screen.  

The approach I took is based on the WndProc subclass process in https://github.com/cnbluefire/WinUI3TransparentBackground.  Rather than always drawing Black, I draw one thing during the startup process and White after that.  I will enhance this soon to draw the SplashScreen image configured in the csproj.

I put this class in a separate assembly (to enable unsafe compilation):

```cs
    public class WinSplash
    {
        private ComCtl32.SUBCLASSPROC wndProcHandler;

        public bool UxInitialized { get; set; }

        public WinSplash(Microsoft.UI.Xaml.Window window)
        {
            var windowHandle = window.GetWindowHandle();

            this.wndProcHandler = new ComCtl32.SUBCLASSPROC(this.WndProc);
            ComCtl32.SetWindowSubclass(windowHandle, this.wndProcHandler, 1, IntPtr.Zero);
        }

        private unsafe IntPtr WndProc(HWND hWnd, uint uMsg, IntPtr wParam, IntPtr lParam, nuint uIdSubclass, IntPtr dwRefData)
        {
            if (uMsg == (uint)User32.WindowMessage.WM_PAINT)
            {
                var hdc = User32.BeginPaint(hWnd, out var ps);
                if (hdc.IsNull) return new IntPtr(0);

                HGDIOBJ brush;
                if (this.UxInitialized)
                {
                    brush = Gdi32.GetStockObject(Gdi32.StockObjectType.WHITE_BRUSH);
                }
                else
                {
                    brush = Gdi32.CreateSolidBrush(System.Drawing.Color.FromArgb(214, 91, 75));
                }
                User32.FillRect(hdc, ps.rcPaint, brush);
                return new IntPtr(1);
            }

            return ComCtl32.DefSubclassProc(hWnd, uMsg, wParam, lParam);
        }
    }
```

And then create it in the OnWindowCreated hook: 

`this.winsplash = new WinSplash(window);`

and then in the OnLaunched hook, I set `this.winsplash.UxInitialized = true`.

I also prefer the original Windows title bar and so set:

`window.AppWindow.TitleBar.ExtendsContentIntoTitleBar = false;`

and the colors of the TitleBar to match the splash screen and later the app colors.

If you get this far and try the above, you will notice a flash of light blue which is the standard Maui App Title Bar.  I turned that off by calling `WindowRootView.UpdateAppTitleBar(0, false);`.  I can share that if desired.
